### PR TITLE
Revive GHC 7.4 support and revise the GHC compatibility window for the Cabal library

### LINKED ISSUE #3838: Revise forward/backward compatiblity window for Cabal versions >= 2
Our policy regarding `Cabal` appears to be

> Our GHC support window is three years: that is, the Cabal library must be buildable out-of-the-box with the dependencies that shipped with GHC for at least three years.

However, given that `Cabal` is a special case to `cabal` for custom build-type packages, this compatibility window is too short and results in artificial constraints for install solving. So we should strive to support wide compatibility windows as long as possible in a best effort way, specifically we shouldn't drop support for older GHCs just for mere convenience if there's no well motivated technical reason for doing so.

Quoting @dcoutts from #3833:

> There may be a good argument now for us reconsidering and expanding our support window. With the advent of setup-deps we can have a packages that need older versions of Cabal, and conversely using a new cabal-install with an older ghc may sill involve using packages that require a more recent Cabal. 
>
> So we may have to expand our support window in both directions: point releases of older Cabal to work with newer ghc, and keeping latest Cabal building with older ghc.

### PR DESCRIPTION
**Pull request by 23Skidoo**
Fixes #3838. See discussion in #3838 for rationale.

/cc @hvr @ttuegel

### CODE REVIEW

**Comment by 23Skidoo** [comment]
tl; dr: Cabal's GHC compatibility window is now 5 years instead of 3, cabal-install's compatibility window is still 3 years.

File: .travis.yml [hunk 1]
```diff
@@ -13,6 +13,9 @@ matrix:
      os: linux
    # These don't have -dyn/-prof whitelisted yet, so we have to
    # do the old-style installation
+   - env: GHCVER=7.4.2 SCRIPT=script CABAL_LIB_ONLY=YES
+     os: linux
+     sudo: required
    - env: GHCVER=7.6.3 SCRIPT=script
```
Comment (ttuegel) [inline .travis.yml]: We discussed setting hard compatibility limits. Should cabal-install also stop supporting GHC 7.6?

**Comment by ezyang** [comment]
Does this mean that we will support using GHCs that are up to five years old?

On September 16, 2016 7:40:15 PM GMT+09:00, Mikhail Glushenkov <notifications@github.com> wrote:
>tl; dr: Cabal's GHC compatibility window is now 5 years instead of 3,
>cabal-install's compatibility window is still 3 years.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/haskell/cabal/pull/3846#issuecomment-247570502

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by 23Skidoo** [comment]
> Does this mean that we will stop supporting GHCs via `--with-compiler` that are older than the five year window?

No, that's a separate issue. We currently don't have any written policy on that.

**Comment by ezyang** [comment]
The reason why we extended Cabal library's support window was because when you say `cabal -w ghc-7.4`, when there is a Custom script, we need to be able to build Cabal library with 7.4. So if you don't synchronize the support window for supported `-w` flags and the support window for building Cabal, you end up in the half-half state where in GHC 7.4 you can build any packages, EXCEPT custom ones. Ick.

 For what it's worth, I don't think there are any good reason to force setup scripts to be built with the same version of GHC that the library is going to be built in. There's no link-time relationship. True, the solver has to be able to solve for multiple versions of GHC simultaneously. Is that so bad? Well, I'm certainly not going to implement it!

File: .travis.yml (hunk 1 above)
Comment (23Skidoo) [inline .travis.yml]: In principle we can, but I don't think we should unless there's a valid reason.

**Comment by 23Skidoo** [comment]
> For what it's worth, I don't think there are any good reason to force setup scripts to be built with the same version of GHC that the library is going to be built in. 

You're assuming that everyone has multiple GHCs installed. Even if this were the case, relaxing this assumption is more trouble than it's worth IMO.

**Comment by 23Skidoo** [comment]
> So if you don't synchronize the support window for supported `-w` flags and the support window for building Cabal, you end up in the half-half state where in GHC 7.4 you can build any packages, EXCEPT custom ones. Ick.

Right now the `-w` compatibility window is larger, so this problem shouldn't happen in practice, but we should re-enable `TEST_OLDER` and actually test that it doesn't.

File: README.md [hunk 2]
```diff
@@ -152,15 +152,16 @@ Conventions
...
-* Our GHC support window is three years: that is, the Cabal library
-  must be buildable out-of-the-box with the dependencies that shipped
-  with GHC for at least three years.  The Travis CI checks this, so
-  most developers submit a PR to see if their code works on all
-  these versions of Haskell.  cabal-install must also be buildable
-  on all these GHCs, although it does not have to be buildable
+* Our GHC support window is five years for the Cabal library and three
+  years for cabal-install: that is, the Cabal library must be
+  buildable out-of-the-box with the dependencies that shipped with GHC
+  for at least five years.  The Travis CI checks this, so most
```
Comment (dcoutts) [inline README.md]: Note that once new-build is the default, we may be able to move to a policy where the Cabal lib merely needs to be "buildable on all supported GHCs" rather than the stronger "buildable out-of-the-box with the dependencies that shipped with GHC". That's because new-build will quite happily install new Cabal lib instances to satisfy setup deps and without causing conflicts.

Comment (BardurArantsson) [inline README.md]: I'm beginning to like this new-build thing better and better... :)
